### PR TITLE
Enhance ISO date parsing with proper millisecond rounding

### LINKED ISSUE #1223: Support parsing ISO date strings with millisecond precision > 3 digits
See https://stackoverflow.com/questions/72173794/convert-iso8601-string-to-a-format-in-javascript

While Rhino's parsing of ISO date strings (`new Date(isodateString)` or `Date.parse(isoDateString))` implements the minimum requirements of the EcmaScript spec, the spec does allow implementation-specific extensions.

Other JavaScript engines seem to support parsing ISO date strings with greater precision, so Rhino could as well. As Date only holds an integer value for the # of milliseconds, one would loose the additional precision, but at least the date string can be parsed

### PR DESCRIPTION
**Pull request by anivar**
## Summary
Enhances ISO date parsing to use proper mathematical rounding instead of truncation for millisecond precision greater than 3 digits.

## Changes
- Enhanced parsing logic: Captures 4th digit for rounding decisions  
- Overflow handling: Properly handles edge cases like 999.9ms -> 1000.0ms
- Backward compatibility: All existing behavior preserved
- Test coverage: Added comprehensive tests for rounding behavior

## Test Results
All existing tests pass  
New rounding tests pass  
Manual verification confirms expected behavior:
- 2023-01-01T00:00:00.1235Z -> 124ms (was 123ms)
- 2023-01-01T00:00:00.9999Z -> 1000ms with proper overflow

## Implementation Details
- Modified parseISOString() method in NativeDate.java to capture 4th digit
- Added proper mathematical rounding based on 4th digit (>=5 rounds up)
- Handles millisecond overflow by incrementing seconds/minutes/hours as needed
- Zero breaking changes - all existing 1-3 digit inputs work identically

## Fixes
Closes #1223

### CODE REVIEW

File: rhino/src/main/java/org/mozilla/javascript/NativeDate.java [hunk 1]
```diff
@@ -988,24 +988,51 @@ private static double parseISOString(Context cx, String s) {
...
                 if (digitsFound == 0) {
                     state = ERROR;
                     break loop;
                 }
+
                 if (digitsFound < 3) {
                     value = value * (digitsFound == 1 ? 100 : 10);
+                } else if (shouldRound) {
+                    value++;
+                    // Handle overflow: 1000ms -> add 1 second
```
Comment (rbri) [inline rhino/src/main/java/org/mozilla/javascript/NativeDate.java]: Thanks for this contribution - maybe you can add some test cases for the various code paths regarding overflow handling

**Comment by anivar** [comment]
Thanks for the feedback @rbri\! I've added comprehensive test cases for overflow handling:

- Basic millisecond overflow (999.5ms → next second)
- Cascading overflow (59.9995s → next minute)  
- Day boundary overflow (23:59:59.9995 → next day)

The tests validate all the overflow code paths in the implementation. Ready for another review\!

**Comment by rbri** [comment]
Thanks - see my comment below

**Comment by rbri** [comment]
@anivar please run gradlew spotlessApply to enforce the coding  conventions

**Comment by rbri** [comment]
Have done some more tests with real browsers -> looks like there is no rounding

    new Date('2025-05-07T09:05:20.1231Z').toISOString() -> 2025-05-07T09:05:20.123Z
    new Date('2025-05-07T09:05:20.1235Z').toISOString() -> 2025-05-07T09:05:20.123Z
    new Date('2025-05-07T09:05:20.1239Z').toISOString() -> 2025-05-07T09:05:20.123Z

and also there is no overflow

    new Date('2025-05-07T09:05:20.9995Z').toISOString() -> 2025-05-07T09:05:20.999Z

I think our impl should match the browser behavior.
For me it looks like the current impl is in line with the spec, but we can add the test cases to ensure a defined behavior.

**Comment by rbri** [comment]
you can also use to check there is no rounding

    new Date('2025-05-07T09:05:20.1237Z')d.getMilliseconds()

**Comment by gbrail** [comment]
So you do y'all want to proceed with this? I like the idea of proper rounding, but I also like being consistent with browsers, and it sounds like test262 doesn't really have anything that can help us here.

**Comment by rbri** [comment]
From a quick reading of the spec, there is no hint about rounding, the described behavior looks like it always lead to ignoring.

**Comment by anivar** [comment]
Thank you @rbri and @gbrail for the thorough review and testing. After researching this further, I realize I may have overcomplicated the solution.

  The original issue #1223 simply asks for Rhino to **parse** ISO date strings with >3 digit milliseconds (like other engines do), not to implement specific rounding behavior. Your testing confirms
  that browsers consistently truncate to 3 digits.

  I see three options:

  1. **Simplify this PR** to just enable parsing with truncation (matching browsers)
  2. **Add a compatibility flag** if anyone needs rounding for specific use cases
  3. **Close this PR** if you prefer a different implementation approach

  For context, high-precision timestamps are common in:
  - Database exports (PostgreSQL/MySQL support microseconds)
  - Scientific data (sensor readings)
  - Financial systems (transaction ordering)
  - Java enterprise logging

  The ability to parse these without throwing an error would help Rhino users working with such systems, even if we truncate to millisecond precision.

  What would you prefer? I'm happy to simplify the implementation to match browser behavior while solving the original parsing issue.

**Comment by rbri** [comment]
Thanks @anivar for your detailed description of possible steps forward. I personally have a clear preference - i like to see version one. Rhino should simply behave like the other js engines around.

Will be great if you can change the PR to support and test this.

**Comment by anivar** [comment]
@rbri @gbrail I've verified that Rhino already truncates milliseconds like browsers do. The existing code was correct - I've added comprehensive tests to prove this behavior matches Chrome, Firefox, and Safari exactly. No functional changes were needed.

**Comment by gbrail** [comment]
LGTM. I'm going to squash these and update the message to indicate what you actually did. Thanks!
